File: AndroidProject/CollectorAgentMind/src/main/java/com/example/collectoragentmind/codelets/perception/ClosestAppleDetector.java
package com.example.collectoragentmind.codelets.perception;

import com.example.collectoragentmind.CollectorAgentMind;
import com.example.collectoragentmind.util.Apple;
import com.example.collectoragentmind.util.Vector2;

import java.util.ArrayList;

import br.unicamp.cst.core.entities.Codelet;
import br.unicamp.cst.core.entities.MemoryObject;

public class ClosestAppleDetector extends Codelet {

    /*
        Scans the list of known apples and updates the
        closestAppleMO accordingly. Stores null in the
        MO if there are no known apples
    */

    private MemoryObject touchMO;
    private MemoryObject knownApplesMO;
    private MemoryObject closestAppleMO;
    private MemoryObject positionMO;

    @Override
    public void accessMemoryObjects() {
        this.knownApplesMO = (MemoryObject) this.getInput("KNOWN_APPLES");
        this.positionMO = (MemoryObject) this.getInput("POSITION");
        this.closestAppleMO = (MemoryObject) this.getOutput("CLOSEST_APPLE");
    }

    @Override
    public void calculateActivation() {

    }

    @Override
    public void proc() {
        ArrayList<Apple> knownApples = (ArrayList<Apple>) knownApplesMO.getI();

        if (knownApples.size() > 0) {
            Vector2 position = (Vector2) positionMO.getI();

            if (position != null) {
                Apple closestApple = knownApples.get(0);
                double minDistance = position.distanceTo(closestApple.position);

                for (int i = 1; i < knownApples.size(); i++) {
                    Apple newApple = knownApples.get(i);
                    double newDistance = position.distanceTo(newApple.position);

                    if (newDistance < minDistance) {
                        closestApple = newApple;
                        minDistance = newDistance;
                    }
                }

                closestAppleMO.setI(closestApple);
            }
        } else {
            closestAppleMO.setI(null);
        }

        CollectorAgentMind.getInstance().updateValue(closestAppleMO.getI().toString(), "CLOSEST_APPLE");
    }
}

File: GodotKotlin/CoreModelTrail/src/main/kotlin/com/example/game/cst/codelets/perception/ClosestAppleDetector.kt
package com.example.game.cst.codelets.perception

import br.unicamp.cst.core.entities.Codelet
import br.unicamp.cst.core.entities.MemoryObject
import com.example.game.environment.Apple
import godot.core.Vector2

class ClosestAppleDetector : Codelet() {
    /*
        Scans the list of known apples and updates the
        closestAppleMO accordingly. Stores null in the
        MO if there are no known apples
    */

    private var knownApplesMO: MemoryObject? = null
    private var closestAppleMO: MemoryObject? = null
    private var positionMO: MemoryObject? = null

    override fun accessMemoryObjects() {
        knownApplesMO = getInput("KNOWN_APPLES") as MemoryObject
        positionMO = getInput("POSITION") as MemoryObject
        closestAppleMO = getOutput("CLOSEST_APPLE") as MemoryObject
    }

    override fun calculateActivation() {}

    override fun proc() {
        val knownApples: ArrayList<Apple> = knownApplesMO!!.i as ArrayList<Apple>
        if (knownApples.size > 0) {
            val position: Vector2 = positionMO!!.i as Vector2
            if (position != null) {
                var closestApple: Apple = knownApples[0]
                var minDistance: Double = position.distanceTo(closestApple.getPos())
                for (apple in knownApples) {
                    val newDistance: Double = position.distanceTo(apple.getPos())
                    if (newDistance < minDistance) {
                        closestApple = apple
                        minDistance = newDistance
                    }
                }
                closestAppleMO!!.i = closestApple
            }
        } else {
            closestAppleMO!!.i = null
        }
    }
}